server/schedule/checker/replica_checker.go

```diff
@@ -80,89 +64,101 @@ func (r *ReplicaChecker) Check(region *core.RegionInfo) *operator.Operator {
-		oldPeer, _ := r.selectWorstPeer(region)
-		if oldPeer == nil {
-			checkerCounter.WithLabelValues("replica_checker", "no-worst-peer").Inc()
-			return nil
-		}
-		op, err := operator.CreateRemovePeerOperator("remove-extra-replica", r.cluster, operator.OpReplica, region, oldPeer.GetStoreId())
-		if err != nil {
-			checkerCounter.WithLabelValues("replica_checker", "create-operator-fail").Inc()
-			return nil
-		}
+	if op := r.checkRemoveExtraReplica(region); op != nil {
 		checkerCounter.WithLabelValues("replica_checker", "new-operator").Inc()
 		return op
 	}
-
-	return r.checkBestReplacement(region)
+	if op := r.checkLocationReplacement(region); op != nil {
+		checkerCounter.WithLabelValues("replica_checker", "new-operator").Inc()
+		return op
+	}
+	return nil
 }
 
-// SelectBestReplacementStore returns a store id that to be used to replace the old peer and distinct score.
-func (r *ReplicaChecker) SelectBestReplacementStore(region *core.RegionInfo, oldPeer *metapb.Peer, filters ...filter.Filter) (uint64, float64) {
-	filters = append(filters, filter.NewExcludedFilter(r.name, nil, region.GetStoreIds()))
-	newRegion := region.Clone(core.WithRemoveStorePeer(oldPeer.GetStoreId()))
-	return r.selectBestStoreToAddReplica(newRegion, filters...)
-}
+// selectStoreToAdd returns the store to add a replica to a region.
+// `extraFilters` is used to set up more filters based on the context that
+// calling this method.
+//
+// For example, to select a target store to replace a region's peer, we can
+// first create a tmp region instance that removed the peer, then call
+// `selectStoreToAdd` to decide the target store. Meanwhile, we need to provide
+// more constraints to ensure that the newly selected node cannot be the same
+// as the original one, and the isolation level cannot be reduced after
+// replacement.
+func (r *ReplicaChecker) selectStoreToAdd(region *core.RegionInfo, extraFilters ...filter.Filter) uint64 {
+	// The selection process uses a two-stage fashion. The first stage
+	// ignores the temporary state of the stores and selects the stores
+	// with the highest score according to the location label. The second
+	// stage considers all temporary states and capacity factors to select
+	// the most suitable target.
+	//
+	// The reason for it is to prevent the non-optimal replica placement due
+	// to the short-term state, resulting in redundant scheduling.
 
-// selectBestPeerToAddReplica returns a new peer that to be used to add a replica and distinct score.
-func (r *ReplicaChecker) selectBestPeerToAddReplica(region *core.RegionInfo, filters ...filter.Filter) (*metapb.Peer, float64) {
-	storeID, score := r.selectBestStoreToAddReplica(region, filters...)
-	if storeID == 0 {
-		log.Debug("no best store to add replica", zap.Uint64("region-id", region.GetID()))
-		return nil, 0
+	filters := []filter.Filter{
+		filter.NewExcludedFilter(replicaCheckerName, nil, region.GetStoreIds()),
```

Comment:
Should we consider `filter.NewStoreLimitFilter(name)`?

Reply:
Included in `StoreStateFilter`.